op-node/rollup: Promote all attributes to safe post-Holocene

sebastianst opened the pull request:
**Description**

Holocene removes backwards-invalidation by span batches. So all derived attributes can immediately be promoted to **safe**. This is the only change in logic of this PR.

The PR also renames fields `IsLastInSpan` to `Concluding` in many structs because that's semantically clearer. That a last batch in a span batch is pre-Holocene the indicator for a safe block promotion is an implementation detail. It's better to name the *effect* rather than the *reason*. Concluding was chosen, since these attributes conclude the pending phase.

**Tests**

Changed an assertion in a test from using pending safe to safe to test this change. It failed before.

**Additional context**

Holocene skips overlapping span batch checks, which in a constructed malicious sequencer setting can lead to a local chain split of nodes that force-restarted mid-way through block derivation. However, the same situation can already be applied pre-Holocene with singular batches, because they are also fast-dropped and overlapping singular batches are not tested to match an existing safe chain. So we may actually want to change the mechanics of safe chain promotion from pending safe to only promote after a full L1 block has been derived. This is reserved for a follow, more details can be found here: https://github.com/ethereum-optimism/optimism/issues/12695

sebastianst commented on op-e2e/actions/upgrades/holocene_fork_test.go:
```diff
@@ -191,8 +191,7 @@ func TestHoloceneInvalidPayload(gt *testing.T) {
 
-	// TODO(12695): need to properly update safe after completed L1 block derivation
-	l2Safe := env.Seq.L2PendingSafe()
+	l2Safe := env.Seq.L2Safe()
```
This is the fixed test.

sebastianst commented on op-node/rollup/derive/batch_stage.go:
```diff
@@ -81,7 +81,7 @@ func (bs *BatchStage) NextBatch(ctx context.Context, parent eth.L2BlockRef) (*Si
 	case BatchAccept: // continue
 		batch.LogContext(bs.Log()).Debug("Found next singular batch")
-		return batch, len(bs.nextSpan) == 0, nil
+		return batch, true, nil // in Holocene, blocks immediately become safe
```
This is the only logic change of this PR

protolambda replied:
Wait, where is the `isHolocene` check? Doesn't this make all pre-holocene batches appear as safe too, while they should be pending-safe?

sebastianst replied:
The `BatchStage` is only active with Holocene, before Holocene, the old `BatchQueue` is active. Stage multiplexing in the derivation pipeline got introduced in https://github.com/ethereum-optimism/optimism/pull/12506

tynes replied:
It could be good to add this context as a comment

sebastianst replied:
improved the comment

sebastianst reviewed (commented):
Marked the actual changes. Everything else is just renaming `IsLastInSpan` to `Safe`.

protolambda reviewed (commented):
> The PR also renames fields IsLastInSpan to Safe in many structs because that's semantically clearer. That a last batch in a span batch is pre-Holocene the indicator for a safe block promotion is an implementation detail. It's better to name the effect rather than the reason. And in some deriver events, it was already called Safe.

While I generally agree with the effect vs reason naming, unfortunately it does not work so well here:
- I find "safe" to be confusing with the "pending safe" concept.
- "safe" (as RPC label) is equal to cross-safe in interop, whereas deriving it for the chain without cross-L2 checks is only considered "local safe" in interop, and thus shouldn't be confused with "safe"

And so since the functionality is actively changing, the effect is not the same across forks, and thus I much prefer the naming based on the reason.

Alternatively, we can name it "stableDerived" or maybe just "stable", or something that makes it clear it's not a pending attribute (nor a sequencer unsafe attribute). (Named after the original 'stable batch derivation' feature name)

sebastianst committed:
op-node/rollup: Promote all attributes to safe post-Holocene

Also renames fields `IsLastInSpan` to `Safe` because that's semantically
clearer. That a last batch in a span batch is pre-Holocene the indicator
for a safe block promotion is an implementation detail. It's better to
name the effect rather than the reason.
